update emojis

porterjamesj opened the pull request:
Update emoji to the latest google emoji set.

This involves:

- get the latest version of `NotoColorEmoji.ttf`, replace the old one with it.
- update `emoji_map.json` accordingly.
- update `dump_emoji.py` to account for the less useful filenames that `ttx` produces from the new `NotoColorEmoji.ttf`. They're of the form `glyph00023.png`, and the numbers have no relation to the unicode codepoint. Happily, the `NotoColorEmoji.ttx` that `ttx` produces has enough information that we can figure out what file is what emoji. `dump_emoji.py` now parses this info and uses to to decide where to put things. I also made some misc cleanups of this script while I was at it.
- update `emoji.js` so the completer knows about the new emoji

@timabbott?

smarx commented:
*Automated message from Dropbox CLA bot*

@porterjamesj, thanks for the pull request! It looks like you haven't yet signed the Dropbox CLA. Please [sign it here](https://opensource.dropbox.com/cla/).

porterjamesj commented:
(signed the CLA)

smarx commented:
*Automated message from Dropbox CLA bot*
				
@porterjamesj, thanks for signing the CLA!

timabbott commented:
Awesome, thanks for doing this @porterjamesj!  

I get this traceback when I try running this:
```
(zulip-venv) tabbott@zaset:~/zulip$ ./tools/setup/emoji_dump/build_emoji 
+ rm -rf static/third/gemoji/images
+ cd tools/setup/emoji_dump
+ python ./emoji_dump.py
Dumping "NotoColorEmoji.ttf" to "NotoColorEmoji#1.ttx"...
Dumping 'GlyphOrder' table...  (22:00:27)
Reading 'post' table from disk  (22:00:27)
Decompiling 'post' table  (22:00:27)
Reading 'maxp' table from disk  (22:00:27)
Decompiling 'maxp' table  (22:00:27)
Reading 'cmap' table from disk  (22:00:27)
Decompiling 'cmap' table  (22:00:27)
Reading 'head' table from disk  (22:00:27)
Decompiling 'head' table  (22:00:27)
Dumping 'head' table...  (22:00:27)
Reading 'hhea' table from disk  (22:00:27)
Decompiling 'hhea' table  (22:00:27)
Dumping 'hhea' table...  (22:00:27)
Dumping 'maxp' table...  (22:00:27)
Reading 'OS/2' table from disk  (22:00:27)
Decompiling 'OS/2' table  (22:00:27)
Dumping 'OS/2' table...  (22:00:27)
Reading 'hmtx' table from disk  (22:00:27)
Decompiling 'hmtx' table  (22:00:27)
Dumping 'hmtx' table...  (22:00:27)
Dumping 'cmap' table...  (22:00:27)
Reading 'name' table from disk  (22:00:27)
Decompiling 'name' table  (22:00:27)
Dumping 'name' table...  (22:00:27)
Dumping 'post' table...  (22:00:27)
Reading 'CBDT' table from disk  (22:00:27)
Decompiling 'CBDT' table  (22:00:27)
Reading 'CBLC' table from disk  (22:00:27)
Decompiling 'CBLC' table  (22:00:27)
Dumping 'CBDT' table...  (22:00:27)
Dumping 'CBLC' table...  (22:00:27)
Reading 'GSUB' table from disk  (22:00:27)
Decompiling 'GSUB' table  (22:00:27)
Dumping 'GSUB' table...  (22:00:27)
Done dumping TTX  (22:00:27)
Traceback (most recent call last):
  File "./emoji_dump.py", line 110, in <module>
    main()
  File "./emoji_dump.py", line 89, in main
    color_font(code_point, code_point_to_fname_map)
  File "./emoji_dump.py", line 20, in color_font
    code_point_to_fname_map[int(code_point, 16)]
KeyError: 52
```

The test failure is a mypy issue; we can investigate that together tomorrow (I think it might be an issue with the `xml` module stubs in typeshed)

porterjamesj commented:
hmm, `Dumping "NotoColorEmoji.ttf" to "NotoColorEmoji#1.ttx"...` suggests to me that you might have an old version of `NotoColorEmoji.ttx` lying around, generated from the old version of `NotoColorEmoji.ttf`. I changed the script to be aggressive about deleting anything like this before it runs, which seems like a good idea even if it's not the cause of the problem you're having.

timabbott commented:
OK, cool, it runs for me now

timabbott commented:
Do you have a diff to the list of valid emoji coming out of this?

porterjamesj commented:
removed emoji:

```
>   "black_square",
>   "white_square",
```

added emoji:

```
<   "a_button",
<   "ab_button",
<   "admission_tickets",
<   "airplane_arrival",
<   "airplane_departure",
<   "alien_monster",
<   "american_football",
<   "amphora",
<   "anger_symbol",
<   "angry_face",
<   "anguished_face",
<   "antenna_bars",
<   "anticlockwise_arrows_button",
<   "artist_palette",
<   "astonished_face",
<   "atm_sign",
<   "atom_symbol",
<   "automobile",
<   "b_button",
<   "baby_angel",
<   "back_arrow",
<   "backhand_index_pointing_down",
<   "backhand_index_pointing_left",
<   "backhand_index_pointing_right",
<   "backhand_index_pointing_up",
<   "badminton",
<   "ballot_box_with_ballot",
<   "barber_pole",
<   "beach_with_umbrella",
<   "bear_face",
<   "beating_heart",
<   "bed",
<   "beer_mug",
<   "bell_with_slash",
<   "bellhop_bell",
<   "bento_box",
<   "bicycle",
<   "billiards",
<   "birthday_cake",
<   "blue_circle",
<   "bottle_with_popping_cork",
<   "bow_and_arrow",
<   "bright_button",
<   "building_construction",
<   "burrito",
<   "bus_stop",
<   "camera_with_flash",
<   "camping",
<   "candle",
<   "card_file_box",
<   "card_index_dividers",
<   "carp_streamer",
<   "castle",
<   "cat_face",
<   "cat_face_with_tears_of_joy",
<   "cat_face_with_wry_smile",
<   "chains",
<   "chart_decreasing",
<   "chart_increasing",
<   "chart_increasing_with_yen",
<   "cheese_wedge",
<   "chequered_flag",
<   "chipmunk",
<   "circled_accept_ideograph",
<   "circled_advantage_ideograph",
<   "circled_letter_m",
<   "cityscape",
<   "cityscape_at_dusk",
<   "clapper_board",
<   "clapping_hands",
<   "classical_building",
<   "clinking_beer_mugs",
<   "clockwise_vertical_arrows",
<   "closed_mailbox_with_lowered_flag",
<   "closed_mailbox_with_raised_flag",
<   "cloud_with_lightning",
<   "cloud_with_lightning_and_rain",
<   "cloud_with_rain",
<   "cloud_with_snow",
<   "cocktail_glass",
<   "coffin",
<   "compression",
<   "computer_mouse",
<   "confounded_face",
<   "confused_face",
<   "control_knobs",
<   "cooked_rice",
<   "cooking",
<   "couch_and_lamp",
<   "cow_face",
<   "crab",
<   "crayon",
<   "cricket",
<   "cross_mark",
<   "cross_mark_button",
<   "crying_face",
<   "curry_rice",
<   "dagger",
<   "dashing",
<   "delivery_truck",
<   "derelict_house_building",
<   "desert",
<   "desert_island",
<   "desktop_computer",
<   "detective",
<   "diamond_with_a_dot",
<   "dim_button",
<   "direct_hit",
<   "disappointed_but_relieved_face",
<   "disappointed_face",
<   "dog_face",
<   "dollar_banknote",
<   "dotted_six_pointed_star",
<   "double_curly_loop",
<   "double_exclamation_mark",
<   "dove",
<   "down_arrow",
<   "down_button",
<   "e_mail",
<   "ear_of_corn",
<   "eight_oclock",
<   "eight_thirty",
<   "eject_button",
<   "eleven_oclock",
<   "eleven_thirty",
<   "end_arrow",
<   "euro_banknote",
<   "evergreen",
<   "expressionless_face",
<   "eye",
<   "face_massage",
<   "face_savouring_delicious_food",
<   "face_screaming_in_fear",
<   "face_throwing_a_kiss",
<   "face_with_cold_sweat",
<   "face_with_head_bandage",
<   "face_with_medical_mask",
<   "face_with_open_mouth",
<   "face_with_open_mouth_and_cold_sweat",
<   "face_with_rolling_eyes",
<   "face_with_steam_from_nose",
<   "face_with_stuck_out_tongue",
<   "face_with_stuck_out_tongue_and_tightly_closed_eyes",
<   "face_with_stuck_out_tongue_and_winking_eye",
<   "face_with_tears_of_joy",
<   "face_with_thermometer",
<   "face_without_mouth",
<   "fast_down_button",
<   "fast_forword_button",
<   "fast_reverse_button",
<   "fast_up_button",
<   "fax_machine",
<   "fearful_face",
<   "ferry",
<   "field_hockey",
<   "file_cabinet",
<   "film_frames",
<   "film_projector",
<   "fish_cake_with_swirl",
<   "fishing_pole",
<   "five_oclock",
<   "five_thirty",
<   "flag_in_hole",
<   "fleur_de_lis",
<   "flexed_biceps",
<   "flushed_face",
<   "fog",
<   "folded_hands",
<   "fork_and_knife_with_plate",
<   "fountain_pen",
<   "four_oclock",
<   "four_thirty",
<   "frame_with_picture",
<   "french_fries",
<   "frog_face",
<   "front_facing_baby_chick",
<   "frowning_face_with_open_mouth",
<   "fuel_pump",
<   "funeral_urn",
<   "gem_stone",
<   "gesturing_no",
<   "gesturing_ok",
<   "glasses",
<   "globe_showing_americas",
<   "globe_showing_asia_australia",
<   "globe_showing_europe_africa",
<   "glowing_star",
<   "goblin",
<   "golfer",
<   "graduation_cap",
<   "grimacing_face",
<   "grinning_cat_face_with_smiling_eyes",
<   "grinning_face",
<   "grinning_face_with_smiling_eyes",
<   "growing_heart",
<   "hammer_and_wrench",
<   "hamster_face",
<   "happy_person_raising_hand",
<   "headphone",
<   "heart_with_arrow",
<   "heart_with_ribbon",
<   "heavy_large_circle",
<   "helmet_with_white_cross",
<   "high_heeled_shoe",
<   "high_speed_train",
<   "high_speed_train_with_bullet_nose",
<   "high_voltage",
<   "hole",
<   "horizontal_traffic_light",
<   "horse_face",
<   "hot_dog",
<   "hot_pepper",
<   "hourglass_with_flowing_sand",
<   "house_building",
<   "house_buildings",
<   "hugging_face",
<   "hundred_points",
<   "hushed_face",
<   "ice_hockey_stick_and_puck",
<   "ice_skate",
<   "index_pointing_up",
<   "input_latin_letters",
<   "input_latin_lowercase",
<   "input_latin_uppercase",
<   "input_numbers",
<   "input_symbols",
<   "japanese_dolls",
<   "japanese_post_office",
<   "japanese_symbol_for_beginner",
<   "joker",
<   "joystick",
<   "kaaba",
<   "kiss_mark",
<   "kissing_cat_face_with_closed_eyes",
<   "kissing_face",
<   "kissing_face_with_closed_eyes",
<   "kissing_face_with_smiling_eyes",
<   "kitchen_knife",
<   "knife",
<   "label",
<   "lady_beetle",
<   "laptop_computer",
<   "last_track_button",
<   "latin_cross",
<   "leaf_fluttering_in_wind",
<   "left_arrow",
<   "left_arrow_curving_right",
<   "left_pointing_magnifying_glass",
<   "left_speech_bubble",
<   "level_slider",
<   "light_bulb",
<   "linked_paperclips",
<   "lion_face",
<   "litter_in_bin_sign",
<   "lock_with_pen",
<   "locomotive",
<   "loud_sound",
<   "loudly_crying_face",
<   "mahjong_red_dragon",
<   "man_and_woman_holding_hands",
<   "man_in_business_suit_levitating",
<   "man_with_chinese_cap",
<   "mantelpiece_clock",
<   "map_of_japan",
<   "megaphone",
<   "menorah",
<   "mens_room",
<   "middle_finger",
<   "military_medal",
<   "moai",
<   "mobile_phone",
<   "mobile_phone_with_arrow",
<   "money_bag",
<   "money_mouth_face",
<   "moon_ceremony",
<   "mosque",
<   "motor_boat",
<   "motorcycle",
<   "motorway",
<   "mountain",
<   "mountain_biker",
<   "mouse_face",
<   "mouth",
<   "musical_notes",
<   "nail_polish",
<   "national_park",
<   "nerd_face",
<   "new_moon_face",
<   "next_track_button",
<   "night_with_stars",
<   "nine_oclock",
<   "nine_thirty",
<   "no_littering",
<   "no_one_under_eighteen",
<   "non_potable_water",
<   "o_button",
<   "office_building",
<   "ogre",
<   "oil_drum",
<   "old_key",
<   "old_man",
<   "old_woman",
<   "om",
<   "on_arrow",
<   "oncoming_fist",
<   "one_oclock",
<   "one_thirty",
<   "open_lock",
<   "open_mailbox_with_lowered_flag",
<   "open_mailbox_with_raised_flag",
<   "optical_disc",
<   "p_button",
<   "paintbrush",
<   "party_popper",
<   "passenger_ship",
<   "pause_button",
<   "peace_symbol",
<   "pedestrian",
<   "pen",
<   "pensive_face",
<   "persevering_face",
<   "person_bowing",
<   "person_in_bed",
<   "person_pouting",
<   "person_raising_hands",
<   "person_taking_bath",
<   "person_with_ball",
<   "pick",
<   "pig_face",
<   "pile_of_poo",
<   "pine_decoration",
<   "ping_pong",
<   "pistol",
<   "place_of_worship",
<   "play_button",
<   "play_or_pause_button",
<   "police_cars_light",
<   "police_officer",
<   "popcorn",
<   "pot_of_food",
<   "pound_banknote",
<   "pouting_cat_face",
<   "pouting_face",
<   "prayer_beads",
<   "printer",
<   "prohibited",
<   "rabbit_face",
<   "racing_car",
<   "railway_track",
<   "raised_fist",
<   "raised_hand_with_fingers_splayed",
<   "record_button",
<   "recreational_vehicle",
<   "recycling_symbol",
<   "red_apple",
<   "red_paper_lantern",
<   "red_triangle_pointed_down",
<   "red_triangle_pointed_up",
<   "relieved_face",
<   "reminder_ribbon",
<   "repeat_button",
<   "repeat_single_button",
<   "reverse_button",
<   "right_anger_bubble",
<   "right_arrow",
<   "right_arrow_curving_left",
<   "right_pointing_magnifying_glass",
<   "roasted_sweet_potato",
<   "robot_face",
<   "rolled_up_newspaper",
<   "rosette",
<   "running_shirt",
<   "running_shoe",
<   "santa_claus",
<   "satellite_antenna",
<   "school_backpack",
<   "scorpion",
<   "seven_oclock",
<   "seven_thirty",
<   "sheaf_of_rice",
<   "shield",
<   "shinto_shrine",
<   "shooting_star",
<   "shopping_bags",
<   "shortcake",
<   "shuffle_tracks_button",
<   "sign_of_the_horns",
<   "simple_smile",
<   "six_oclock",
<   "six_thirty",
<   "skier",
<   "skis",
<   "sleeping_face",
<   "sleepy_face",
<   "slightly_frowning_face",
<   "slightly_smiling_face",
<   "small_airplane",
<   "smiling_cat_face_with_heart_shaped_eyes",
<   "smiling_cat_face_with_open_mouth",
<   "smiling_face",
<   "smiling_face_with_halo",
<   "smiling_face_with_heart_shaped_eyes",
<   "smiling_face_with_horns",
<   "smiling_face_with_open_mouth",
<   "smiling_face_with_open_mouth_and_cold_sweat",
<   "smiling_face_with_open_mouth_and_smiling_eyes",
<   "smiling_face_with_open_mouth_and_tightly_closed_eyes",
<   "smiling_face_with_smiling_eyes",
<   "smiling_face_with_sunglasses",
<   "smirking_face",
<   "snow_capped_mountain",
<   "snowman_without_snow",
<   "soccer_ball",
<   "soft_ice_cream",
<   "soon_arrow",
<   "speaker_loud",
<   "speaker_off",
<   "speaker_on",
<   "speaking_head",
<   "spider",
<   "spider_web",
<   "spiral_calendar",
<   "spiral_notepad",
<   "spiral_shell",
<   "sports_medal",
<   "spouting_whale",
<   "squared_apply_ideograph",
<   "squared_cl",
<   "squared_cool",
<   "squared_divide_ideograph",
<   "squared_empty_ideograph",
<   "squared_exist_ideograph",
<   "squared_finger_ideograph",
<   "squared_free",
<   "squared_fullness_ideograph",
<   "squared_id",
<   "squared_katakana_koko",
<   "squared_katakana_sa",
<   "squared_moon_ideograph",
<   "squared_negation_ideograph",
<   "squared_new",
<   "squared_ng",
<   "squared_ok",
<   "squared_operating_ideograph",
<   "squared_prohibit_ideograph",
<   "squared_sos",
<   "squared_together_ideograph",
<   "squared_vs",
<   "stadium",
<   "star_and_crescent",
<   "steaming_bowl",
<   "stop_button",
<   "stopwatch",
<   "studio_microphone",
<   "sun_behind_cloud",
<   "sun_behind_cloud_with_rain",
<   "sun_behind_large_cloud",
<   "sun_behind_small_cloud",
<   "sunset",
<   "sweat_droplets",
<   "synagogue",
<   "t_shirt",
<   "taco",
<   "teacup_without_handle",
<   "tear_off_calendar",
<   "television",
<   "ten_oclock",
<   "ten_thirty",
<   "thermometer",
<   "thinking_face",
<   "three_oclock",
<   "three_thirty",
<   "thumbs_down",
<   "thumbs_up",
<   "tiger_face",
<   "timer_clock",
<   "top_arrow",
<   "top_hat",
<   "tornado",
<   "trackball",
<   "tram_car",
<   "triangular_flag",
<   "trident_emblem",
<   "turkey",
<   "twelve_oclock",
<   "twelve_thirty",
<   "two_hump_camel",
<   "two_oclock",
<   "two_thirty",
<   "umbrella_on_ground",
<   "umbrella_with_rain_drops",
<   "unamused_face",
<   "unicorn_face",
<   "up_arrow",
<   "up_button",
<   "upside_down_face",
<   "victory_hand",
<   "videocassette",
<   "volleyball",
<   "vulcan_salute",
<   "wastebasket",
<   "water_closet",
<   "water_wave",
<   "waving_black_flag",
<   "waving_hand",
<   "waving_white_flag",
<   "weary_cat_face",
<   "weary_face",
<   "weight_lifter",
<   "white_medium_star",
<   "wind_face",
<   "winking_face",
<   "wolf_face",
<   "womans_boot",
<   "womans_sandal",
<   "women_partying",
<   "womens_room",
<   "world_map",
<   "worried_face",
<   "wrapped_present",
<   "writing_hand",
<   "yen_banknote",
<   "yin_yang",
<   "zipper_mouth_face",
```

some of the old emojis also have different images

porterjamesj commented:
^ some of those are just new names for preexisting emoji, not sure how much we care about that.

timabbott commented:
Awesome, people are going to enjoy all those new emoji! :)

timabbott commented:
Can you split out the reformatting part of https://github.com/zulip/zulip/pull/875/commits/e12c6e6f14a5d54179c1c2174ae7b16266eb136e into its own commit?  I think it'll be nice for folks reading the history to be able to see just the functional changes, which seems like it'd be like 1/3 of the diff, and I think it should be pretty fast to do.  

(and we'll need to debug the mypy thing, otherwise this looks ready to merge!)

porterjamesj commented:
o7

porterjamesj commented:
Confirmed that fixing the stubs in typeshed fixes the mypy error above. PR fixing this: https://github.com/python/typeshed/pull/254

We can either wait for that PR to make it's way into mypy and upgrade mypy before merging this, or I can just do a hack (use `.iter`, use `.getchildren()`, etc.) to get mypy to be happy and we can merge sooner. I'm happy to do either.

timabbott commented:
Cool.  I think it probably makes sense to use a `# type: ignore # https://github.com/python/typeshed/pull/254` to suppress the mypy issue while we wait for your fix to be merged

timabbott commented:
Emoji waits for no man, so I went ahead and did that tweaks, and merged this, after adjusting capitalization in your commit messages to match our style (should probably add a linter for that).

Thanks @porterjamesj for doing this!!!

porterjamesj commented:
Thanks! Sorry for delay, I was traveling and moving for the last couple days.

timabbott commented:
No worries, I'm just excited about our new modern emoji.  I would celebrate with :upside_down_face:, but apparently GitHub doesn't have it :frowning:
